client: bond rotation

chappjc opened the pull request:
Putting up this draft PR before it's ready to help inform https://github.com/decred/dcrdex/pull/2025. This draft PR includes the main bond lifecycle machinery for the client.  I've not pushed a commit for the reserves, except for a comment in client/asset/interface.go.

---

**EDIT:**  We can review reserves in a second PR.  I have experimented with two different approaches, each with their own pros/cons:
- explicit UTXO-based reserves, using lower level coin locking, and requires making transactions to create the UTXO (and replenish as network fees reduce it over time)
- virtual balance based reserves, using higher level balance and available/spendable checks and adjustments, avoiding coin locking and need to make a reserve UTXO, but it becomes hairy to prevent accidentally spending the reserve UTXO

I really want the virtual reserves to work because the extra transaction creation of UTXO-based reserves has on-chain overhead and requires maintenance as fees eat into this reserve, whereas the virtual reserves is just an amount that is continually respected when funding transactions or orders.  Personally I think we want a "softer" kind of reserves like the virtual balance based-approach, although it has more edges.

---

```
// Bond lifetime
//
//   t0  t1                      t2'    t2                   t3  t4
//   |~~~|-----------------------^------|====================|+++|
//     ~                 -                        =
//  pending (p)       live (l)                expired (E)   maturing (m)
//
//    t0 = authoring/broadcast
//    t1 = activation (confirmed and accepted)
//    t2 = expiry (tier accounting)
//    t3 = lockTime
//    t4 = spendable (block medianTime > lockTime)
//
//    E = t3 - t2, *constant* duration, dex.BondExpiry()
//    p = t1 - t0, *variable*, a random process
//    m = t4 - t3, *variable*, depends on consensus rules and blocks
//        e.g. several blocks after lockTime passes
//
//  - bonds may be spent at t4
//  - bonds must be replaced by t2 i.e. broadcast by some t2'
//  - perfectly aligning t4 for bond A with t2' for bond B is impossible on
//    account of the variable durations
//  - t2-t2' should be greater than a large percent of expected pending
//    durations (see pendingBuffer)
//
// Here a replacement bond B had a long pending period, and it became active
// after bond A expired (too late):
//
//             t0  t1                         t2' t2                   t3
//   bond A:   |~~~|--------------------------^---|====================|
//                                                x
//   bond B:                                  |~~~~~~|------------------...
//
// Here the replacement bond was confirmed quickly, but l was too short,
// causing it to expire before bond A became spendable:
//                                                                        > renew as bond C
//   bond A:   |~~~|----------------------^-------|====================|++‖~~~~~|---...
//                                          ✓                             x     x
//   bond B:                              |~|------------------------|=====...
//
// Similarly, l could have been long enough to broadcast a replacement in time,
// but the pending period could be too long (the second "x").
//
// Here the replacement bond was broadcast with enough time to confirm before
// the previous bond expired, and the previous bond became spendable in time to
// broadcast and confirm another replacement (sustainable):
//                                                                        > renew as bond C
//   bond A:   |~~~|----------------------^-------|====================|++‖~~~~~|---...
//                                               ✓                        ✓     ✓
//   bond B:                              |~~~~~~|-------------------------------|====...
//
// Thus, bond rotation without tier drops requires l>E+m+p. For
// t3-t0 = p+l+E, this means t3-t0 >= 2(E+p)+m. We will assume the time
// from authoring to broadcast is negligible, absorbed into the estimate of the
// max pending duration.
//
// tldr:
//  - when creating a bond, set lockTime = 2*(BondExpiry+pendingBuffer)+spendableDelay
//  - create a replacement bond at lockTime-BondExpiry-pendingBuffer
```

The `client/core` form types:

```go
// BondOptionsForm is used from the settings page to change the auto-bond
// maintenance setting for a DEX.
type BondOptionsForm struct {
	Addr         string  `json:"host"`
	TargetTier   *uint64 `json:"targetTier,omitempty"`
	MaxBondedAmt *uint64 `json:"maxBondedAmt,omitempty"`
	BondAsset    *uint32 `json:"bondAsset,omitempty"`
}

// PostBondForm is information necessary to post a new bond for a new or
// existing DEX account at the specified DEX address.
type PostBondForm struct {
	Addr     string           `json:"host"`
	AppPass  encode.PassBytes `json:"appPass"`
	Asset    *uint32          `json:"assetID,omitempty"` // do not default to 0
	Bond     uint64           `json:"bond"`
	LockTime uint64           `json:"lockTime"` // 0 means go with server-derived value

	// These options may be set when creating an account.
	MaintainTier *bool   `json:"maintainTier,omitempty"` // tier implied from Bond amount
	MaxBondedAmt *uint64 `json:"maxBondedAmt,omitempty"`

	// Cert is needed if posting bond to a new DEX. Cert can be a string, which
	// is interpreted as a filepath, or a []byte, which is interpreted as the
	// file contents of the certificate.
	Cert interface{} `json:"cert"`
}
```

norwnd commented on client/cmd/dexcctl/dexcctl_test.go:
```diff
@@ -75,20 +75,20 @@ func TestConfigure(t *testing.T) {
 	os.Args = []string{"", "-7"}
 	_, _, _, err = configure()
 	if err == nil {
-		t.Fatal("expected failure on bad flag")
+		t.Fatal("expected success on bad flag")
```
`expected success` here doesn't add up with `if err == nil` check from above ? Not sure what's the intention, just caught my eye.

chappjc commented:
I have tested at length on testnet, and have ironed out most edges that pop out with the unreliable dextest hidden service.  Making ready for review now without the reserves since (a) this is a good deal to review on it's own, (b) we need these API changes for https://github.com/decred/dcrdex/pull/2025, and (c) I'm wavering on the reserves solution and need more time, see below.

We can review reserves in a second PR.  I have experimented with two different approaches, each with their own pros/cons:
- explicit UTXO-based reserves, using lower level coin locking, and requires making transactions to create the UTXO (and replenish as network fees reduce it over time)
- virtual balance based reserves, using higher level balance and available/spendable checks and adjustments, avoiding coin locking and need to make a reserve UTXO, but it becomes hairy to prevent accidentally spending the reserve UTXO

I really want the virtual reserves to work because the extra transaction creation of UTXO-based reserves has on-chain overhead and requires maintenance as fees eat into this reserve, whereas the virtual reserves is just an amount that is continually respected when funding transactions or orders.  Personally I think we want a "softer" kind of reserves like the virtual balance based-approach, although it has more edges.

chappjc commented on client/asset/dcr/rpcwallet.go:
```diff
@@ -525,6 +525,17 @@ func (w *rpcWallet) InternalAddress(ctx context.Context, acctName string) (stdad
 }
 
+// ExternalAddress returns an external address from the specified account using
+// GapPolicyWrap. The dcrwallet user should set --gaplimit= as needed to prevent
+// address reused depending on their needs. Part of the Wallet interface.
+func (w *rpcWallet) ExternalAddress(ctx context.Context, acctName string) (stdaddr.Address, error) {
+	addr, err := w.rpcClient.GetNewAddressGapPolicy(ctx, acctName, dcrwallet.GapPolicyWrap)
```
As the revised comment states, with an external dcrwallet, the user can pick their `--gaplimit` as desired.  Using ignore with the RPC has created difficult recover situations.  The `InternalAddress` method already uses wrap implicitly (`getrawchangeaddress` does not even expose gap policy).

chappjc replied:
When an address is used, the next (unused) address index becomes the address to which wrapping returns.  The wrap algo just ensures the gap limit isn't exceeded.

chappjc commented on client/core/notification.go:
```diff
@@ -457,7 +457,7 @@ const TopicBalanceUpdated Topic = "BalanceUpdated"
 
 func newBalanceNote(assetID uint32, bal *WalletBalance) *BalanceNote {
 	return &BalanceNote{
-		Notification: db.NewNotification(NoteTypeBalance, TopicBalanceUpdated, "balance updated", "", db.Data),
```
This constructor was an outlier - other `db.Data` notes use empty subject and details, which is the signal for `logNotify` method to skip logging the not before relaying it to consumers. https://github.com/decred/dcrdex/blob/f02dc3d5d2edc1c8437b1b854aaa11051a79d1c0/client/core/notification.go#L48-L49

chappjc commented on client/core/trade.go:
```diff
@@ -1562,6 +1562,7 @@ func (t *trackedTrade) isRedeemable(ctx context.Context, match *matchTracker) (r
 				if match.MetaData.Proof.SelfRevoked {
 					return false, false // already self-revoked
 				}
+				// Here we can check to see if this is a redeem we failed to record...
```
This comment pertains to https://github.com/decred/dcrdex/pull/2049#discussion_r1069576868
It's unrelated to bonds, but it snuck in.

martonp commented on client/db/types.go:
```diff
@@ -176,7 +179,10 @@ type AccountInfo struct {
 
-	Bonds []*Bond
+	Bonds        []*Bond
+	TargetTier   uint64 // zero means no bond maintenance (allows negative)
```
allows negative?

chappjc replied:
`TargetTier` set to 0 allows the users effective trading tier (according to server) to drop into negative.  As in if this is set to zero it's not going to try to keep it at zero but rather let it go.

martonp reviewed (commented):
Looks good, haven't tested though.

martonp commented:
> * virtual balance based reserves, using higher level balance and available/spendable checks and adjustments, avoiding coin locking and need to make a reserve UTXO, but it becomes hairy to prevent accidentally spending the reserve UTXO

This will be useful for bot inventory management as well.

chappjc commented on client/core/core.go:
```diff
@@ -2682,6 +2687,42 @@ func (c *Core) walletIsActive(assetID uint32) bool {
+	for _, dc := range c.dexConnections() {
+		bondAsset, targetTier, _ := dc.bondOpts()
+		if bondAsset != assetID {
+			continue
```
Although refunds do not require an unlocked wallet (the private key is stored in the `db.Bond`), it would certainly not be ok to *disable* a wallet that has active bonds to refund.   The `includeLive` check below is intended to check that case, but now I see from your comment that the `continue` above will preclude that check unintentionally.  Will fix.

itswisdomagain commented on client/core/bond.go:
```diff
@@ -123,6 +283,14 @@ func (c *Core) refundExpiredBonds(ctx context.Context) {
 				continue
 			}
 
+			// Here we may either refund or renew the bond depending on target
+			// tier and timing. Direct renewal (refund and post in one) is only
+			// useful if there is insufficient reserves or the client had been
+			// stopped for a while. Normally, a bond becoming spendable will not
+			// coincide with the need to post bond.
+			//
+			// TODO: if mustPost > 0 { wallet.RenewBond(...) }
```
Is it possible to spend multiple expired bonds in one renewal, rather than renew individually?

chappjc replied:
It would be possible yes.  Timing would be even less likely in practice though.  Putting this PR into work on testnet where the lock times are much longer helps get a feel for more real life cadences where there's a fair amount of bond overlap and very little coincidence of <it's time to post> with <it's time to refund>.  The only way to make that happen more readily is to simply not refund until <it's time to post>.  I'm reluctant to leave funds locked in bonds longer than needed, but I'm considering it.

Working through some concepts that will help decide about this in a `bond-reserves` branch for a PR very soon.

itswisdomagain commented on client/core/bond.go:
```diff
@@ -408,23 +622,68 @@ func deriveBondKey(seed []byte, assetID, bondIndex uint32) (*secp256k1.PrivateKe
+	}
+	c.log.Debugf("Bond options for %v: target tier %d, bond asset %d, maxBonded %v",
+		acct.Host, dc.acct.targetTier, dc.acct.bondAsset, acct.MaxBondedAmt)
+	return c.db.UpdateAccountInfo(acct)
```
If updating the db fails, should dc.acct revert to old values? Or not even set the new values until db update succeeds.

itswisdomagain reviewed (changes requested):
Will test next.

chappjc commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+
 	for _, dc := range c.dexConnections() {
+		// time.Now().After(time.Unix(bond.LockTime-dc.cfg.BondExpiry))
 		lockTimeThresh := now // in case dex is down, expire (to refund) when lock time is passed
```
"refund when lock time is passed" is grouped separately because "expire" in this case means when the bond ceases to become active for tier accounting purposes, which happens some duration before lockTime, and that duration is the "expiry" duration

itswisdomagain replied:
"refund when lock time is passed" isn't grouped separately currently, it's currently ...
`expire (to refund) when lock time is passed`

chappjc replied:
ah, ok you are correct!

chappjc commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+// max pending duration.
+//
+// tldr:
+//  - when creating a bond, set lockTime = 2*(BondExpiry+pendingBuffer)+spendableDelay
```
Good points.  I was actually considering writing these as `lockTime - t0 = ...` but didn't want them even longer, but since you are picking up on the correctness of that, I will update.

itswisdomagain commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
-					c.log.Warnf("Expired bond found: %v", coinIDString(bond.AssetID, bond.CoinID))
+					c.log.Infof("Newly expired bond found: %v (%s)", coinIDString(bond.AssetID, bond.CoinID), unbip(bond.AssetID))
 				} else {
+					if int64(bond.LockTime) <= replaceThresh {
+						weak = append(weak, bond) // but still live
+					}
```
The bond may be pending. Does that require special handling?

chappjc replied:
Good question.  For purposes of `mustPost` computation, I believe this comes out correctly:

```go
		tierDeficit := int64(targetTier) - tier
		mustPost := tierDeficit + weakStrength - pendingStrength
```

By putting that pending bond in the weak bucket also (note it falls through and stays in `pendingBonds` too), we negate the pending bond's offset to `mustPost`, which I believe is correct.

Will consider that more however.

buck54321 commented on client/core/core.go:
```diff
@@ -4289,6 +4349,16 @@ func (c *Core) Login(pw []byte) error {
 	defer c.loginMtx.Unlock()
 
 	if !c.loggedIn {
+		// Derive the bond extended key from the seed.
+		seed, err := crypter.Decrypt(creds.EncSeed)
+		if err != nil {
+			return fmt.Errorf("seed decryption error: %w", err)
+		}
+		defer encode.ClearBytes(seed)
+		c.bondXPriv, err = deriveBondXPriv(seed)
+		if err != nil {
+			return fmt.Errorf("GenDeepChild error: %w", err)
+		}
```
Should we not store this encrypted and decrypt it on demand? We could make the encrypted key part of the `PrimaryCredentials`.

chappjc replied:
> decrypt it on demand

Not sure I see how that works.  In this scheme login is the "on demand" action.  When you login, bond maintenance begins, and it needs the keys at any moment.

buck54321 replied:
Oh right.

buck54321 commented on client/core/types.go:
```diff
@@ -170,7 +173,10 @@ type PostBondForm struct {
 	Asset    *uint32          `json:"assetID,omitempty"` // do not default to 0
 	Bond     uint64           `json:"bond"`
```
I kinda wish this was `Tier uint64` instead of `Bond uint64`. I suppose using a bond amount allows us to do some validation that might catch user errors, but specifying (tier, asset) seems easier. The UI will use tier as an input (e.g. `bondStrengthField` in #2025), and it would be better via RPC too, but easier to mess up.

chappjc replied:
This is where the `bondopt` RPC (and `UpdateBondOptions` method would apply).  As noted in your previous comment, postbond should really just be used to register.

chappjc replied:
> It looks like the new `rotateBonds` method will use the server's config value as is. Should it check back with this original value? (Maybe it does and I missed it.)

Only indirectly via the `MaxBondedAmt` set, but I think I've broken this just recently.

buck54321 commented on client/db/types.go:
```diff
@@ -271,6 +282,35 @@ func decodeAccountInfo_v2(pushes [][]byte) (*AccountInfo, error) {
 		EncKeyV2:     v2Key,
 		LegacyEncKey: legacyKeyB,
 		// Bonds decoded by DecodeBond from separate pushes.
+		BondAsset:        defaultBondAsset,
```
Nice that there's no update needed.

buck54321 commented on client/rpcserver/handlers.go:
```diff
@@ -1126,14 +1141,24 @@ Registration is complete after the fee transaction has been confirmed.`,
 	bond (int): The bond amount (in DCR presently).
 	assetID (int): The asset ID with which to pay the fee.
-	lockTime (int): The bond's lockTime as UNIX epoch time (seconds).
-    cert (string): Optional. The TLS certificate path.`,
+    cert (string): Optional. The TLS certificate path. Only applicable when registering.
+	maintain (bool): Optional. Whether to maintain the trading tier established by this bond. Only applicable when registering. (default is true)`,
```
Not new, there's a tab/space thing going on here I think.

chappjc commented on client/asset/interface.go:
```diff
@@ -473,11 +473,16 @@ type Broadcaster interface {
 type Bonder interface {
 	Broadcaster
 
+	// ReserveBondFunds reserves funds for creation of future bonds. It may
+	// create a transaction to pre-size the reserved funds. MakeBondTx will
```
It's no longer correct anyway.  I've abandoned this approach.  Will push `bond-reserves` branch tomorrow.

chappjc commented on client/core/bond.go:
```diff
@@ -477,22 +736,38 @@ func (c *Core) PostBond(form *PostBondForm) (*PostBondResult, error) {
+			return nil, fmt.Errorf("maintain tier and max bonded amount may only be set when registering")
+		}
 	} else {
+		maxBondedAmt := 4 * form.Bond // default
```
> `MaxBondedAmt` is a sort of safety feature?

Correct. Client posts bond to hit target tier, but it's server saying what tier you are, so if server keeps reporting low tier (either because penalization or malice), you'll keep posting bonds without limit.

chappjc commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+//  - when creating a bond, set lockTime = 2*(BondExpiry+pendingBuffer)+spendableDelay
+//  - create a replacement bond at lockTime-BondExpiry-pendingBuffer
+
+// pendingBuffer gives the duration in seconds prior to reaching bond expiry
+// (account, not lockTime) after which a new bond should be posted to avoid
+// account tier falling below target while the replacement bond is pending. The
+// network is a parameter because expected block times are network-dependent,
+// and we require short bond lifetimes on simnet.
```
> Instead of a universal approximation, why not have the `Bonder` specify generous estimates for the asset-specific variables`p` and `m` and use `2p + m` (I think) from above?

Right, I've alluded to this idea in the docs for `spendableDelay` below.  

```
// .... NOTE: to minimize bond overlap, an asset.Bonder method could provide
// this estimate, but it is still very short relative to the entire bond
// lifetime, which is on the order of months.
```

I'm just convincing myself that it would matter enough to justify the API bloat and additional code.  Not yet seeing it, but may yet do this when the reserves is working right.

> To clarify, pendingBuffer is not an approximation for pending (p) = t1 - t0, it's t2 - t2'?

That's correct.  This pertains to the second "tldr" bullet from the giant block comment above:

```
// tldr:
//  - when creating a bond, set lockTime = 2*(BondExpiry+pendingBuffer)+spendableDelay
// - create a replacement bond at lockTime-BondExpiry-pendingBuffer
```

chappjc commented on client/core/bond.go:
```diff
@@ -477,22 +736,38 @@ func (c *Core) PostBond(form *PostBondForm) (*PostBondResult, error) {
 			return nil, newError(acctKeyErr, "acct locked %s (login first)", form.Addr)
 		}
+		if form.MaintainTier != nil || form.MaxBondedAmt != nil {
+			return nil, fmt.Errorf("maintain tier and max bonded amount may only be set when registering")
```
Correct.  This path won't apply to UI users, where we'll either give an appropriate registration dialog with options to maintain tier (if a new account), or we'll have the bond options form where they can start/end maintenance to given levels.

chappjc replied:
Related comment: https://github.com/decred/dcrdex/pull/2036/files/d03b9e095393c6ae1d8692b0b4f55a072b4808c1#r1071503810

BTW, I'm sorry reserves is taking so long to put up.  A lot of things relating to target tier and max bonded amt should become clearer.

chappjc commented on client/asset/interface.go:
```diff
@@ -473,11 +473,16 @@ type Broadcaster interface {
+	MakeBondTx(ver uint16, amt, feeRate uint64, lockTime time.Time, privKey *secp256k1.PrivateKey, acctID []byte) (*Bond, error)
 	// RefundBond will refund the bond given the full bond output details and
 	// private key to spend it.
 	RefundBond(ctx context.Context, ver uint16, coinID, script []byte, amt uint64, privKey *secp256k1.PrivateKey) ([]byte, error)
```
Oh, I should mention that I've made `RefundBond` actually broadcast it internally in `bond-reserves`, rather than return the raw tx that gets send by core.  I am positive that was suggested in the previous PR and I just neglected to apply the change.  That pattern only needs to exist for `MakeBondTx`

itswisdomagain commented on client/core/bond.go:
```diff
@@ -521,19 +799,17 @@ func (c *Core) PostBond(form *PostBondForm) (*PostBondResult, error) {
 
-	lockTime := time.Now().Add(2 * bondValidity).Truncate(time.Second) // default lockTime is double
+	bondValidity := minBondLifetime(c.net, int64(bondExpiry))
+	lockTime := time.Now().Add(bondValidity).Truncate(time.Second)
```
Few lines below this, expireTime is wrongly set to `time.Now().Add(bondValidity)`. And I think that should be `earliestServerExpiry := time.Now().Add(bondExpiry)`? 

Or... if it's the actual time the server will expire the new bond, then it should be?
`expireTime := locktime.Add(-bondExpiry)`?

chappjc replied:
Good catch. This `minBondLifetime` refactoring was incomplete here in `PostBond`.

chappjc replied:
Sorry about breaking that.  Fixed that regression in https://github.com/decred/dcrdex/commit/01a356289c9a5ad134f471ba295699c8a44dbc87.
I had started my rotations going before I squashed in that bad `minBondLifetime` update.

Shall we rebase https://github.com/decred/dcrdex/pull/2025 on top of this PR now?  Or are there still questions regarding interface, types, semantics, etc.?

JoeGruffins reviewed (commented):
The answer to this is probably obvious but I'm not sure so, what is the point of the expiry? Why can't the bond be good up until locktime?

chappjc commented:
> The answer to this is probably obvious but I'm not sure so, what is the point of the expiry? Why can't the bond be good up until locktime?

Say there's 1 minute until they can reclaim their bond, there's virtually no incentive to play nice.  Now, if our bonds were designed to be confiscated on violations, that would be different, however that's a bit out of the question, so we just invalidate them (indirectly via an account's tier)

buck54321 commented:
> I really want the virtual reserves to work

Me too

chappjc commented:
> > I really want the virtual reserves to work
> 
> Me too

Haha. It's working!  Just hard to make it pretty and clean.  Getting really close.  Testing edges and polishing the rough bits.

buck54321 commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
 					// This is generally unexpected because auth, reconnect, or a
 					// bondexpired notification should do this first.
 					dc.acct.expiredBonds = append(dc.acct.expiredBonds, bond)
-					c.log.Warnf("Expired bond found: %v", coinIDString(bond.AssetID, bond.CoinID))
+					c.log.Infof("Newly expired bond found: %v (%s)", coinIDString(bond.AssetID, bond.CoinID), unbip(bond.AssetID))
```
This directly expired bond is not considered in the calculation of `tierDeficit` below. Partially related to my other comment on `acct.tier` being an int filed rather than a function.

chappjc replied:
Right, I hope this will be come much clearly soon, but you get a `bondexpired` notification shortly after (although possibly before if clocks are set badly), which informs us of the server's understanding of our account tier.  Similarly, the `tierchange` route is used when penalization change the account tier.  And for reconnect handling, the auth/`connect` route also reports tier.  Collectively, we never miss a beat in terms of our effective tier (what server things our tier is).

chappjc replied:
As an example:

```
2023-01-29 19:10:13.582 [INF] CORE: Newly expired bond found: 90287fa1e27ec54c03a94389c7bf22fe63eb62b0e408ef2cbf8078a0c5cc945e:0 (dcr)
2023-01-29 19:10:13.582 [DBG] CORE: Expired bond 90287fa1e27ec54c03a94389c7bf22fe63eb62b0e408ef2cbf8078a0c5cc945e:0 (dcr) refundable in about 3m55s.
2023-01-29 19:10:15.599 [INF] CORE: Received bondexpired notification from 127.0.0.1:17273 for account ad5afeb6189a56487703568485ea26a8db4e11ebbbe1d32dccddcff9bf8b5812...
2023-01-29 19:10:15.600 [INF] CORE: notify: |SUCCESS| (bondpost) BondExpired - New tier = 2 (target = 2).
2023-01-29 19:10:25.797 [DBG] CORE[dcr]: tip change: 13924 (5c764064241eee42fd95e2f450fe129cccd5d5e69902411b09b0fa1e56e3093b) => 13925 (4fcef9f65c942496e24ca8e3e02d21ea10dd417af31ab5406f30043ef2820935)
2023-01-29 19:10:25.797 [TRC] CORE: Processing tip change for dcr
2023-01-29 19:10:25.797 [TRC] CORE: asset 42: total = 4000000000, active = 2000000000
2023-01-29 19:10:33.582 [DBG] CORE: Expired bond 90287fa1e27ec54c03a94389c7bf22fe63eb62b0e408ef2cbf8078a0c5cc945e:0 (dcr) refundable in about 3m35s.
```

Sorry for all the typos in my last commment.

buck54321 replied:
Ah. So since the server wouldn't consider the expired bond as part of its `ConnectResult.Tier`, but would still consider the weak bond, we're okay to trust `acct.tier` below in our `tierDeficit`. Do we ever validate that the server is reporting the tier that we expect?

chappjc replied:
> Do we ever validate that the server is reporting the tier that we expect?

For this, I think the missing piece is the violation reporting.  Client software always just does it's best, and generally it's not tracking it's own known violations (e.g. whoops I was offline for a while and I'm pretty sure I messed up on swap X, therefore I expect a -1 tier adjustment).  Not sure we'll ever really do that latter thing, but server's comms should certainly list matches and a specific state at which your account failed.

So to have any hope of functioning, we have to be concerned with how server has classified our account.  A related concept is MaxBondedAmt, because we'll start posting bond as needed (when maintenance is enabled) to keep a target tier.

I suppose ideally we'd have something in client that considers what we believe to be our own conduct state.  That's not unique to bonds though.  With the legacy fee, we just got penalized and client had no judgement on the correctness of that, it was up to the user.  I expect we'll evolve this.

buck54321 reviewed (commented):
Looking good.

Feel free to take anything you find useful from https://github.com/buck54321/dcrdex/compare/01a356289c9a5ad134f471ba295699c8a44dbc87...buck54321:dcrdex:rotate-test

chappjc commented on client/core/types.go:
```diff
@@ -704,7 +710,10 @@ type dexAccount struct {
 	bonds         []*db.Bond // confirmed, and not yet expired
 	expiredBonds  []*db.Bond // expired and needing refund
 	tier          int64      // check instead of isSuspended
```
`tier` is the last known tier that the server/network says we are.  This was established previously, but it will become much more clearer in reserves work I'm prepping to make public.

There's a concept of `bondedTier`, which would be what you are describing - a function of known bonds and their statuses.  Then their is effective tier, which is what server says we are.  Both of these concepts come into play, but this value is the deciding factor as it relates to how we expect server to interact with us.

On a related note, we have future work to do with including more penalty info, at least in the form of a single "tier adjustment" integer for this, but ideally with more description of the violations as server sees them.

chappjc commented on client/core/core.go:
```diff
@@ -4785,6 +4879,20 @@ func (c *Core) wait(coinID []byte, assetID uint32, trigger func() (bool, error),
+func (c *Core) waiting(coinID []byte, assetID uint32) bool {
+	c.waiterMtx.RLock()
+	defer c.waiterMtx.RUnlock()
+	_, found := c.blockWaiters[coinIDString(assetID, coinID)]
```
Not sure I follow.  Which change to `coinIDString`?  I think I'm matching the use in `(*Core).wait`, and this new `waiting` method does seem to work as expected.

buck54321 replied:
Nevermind. I was confused by the change in call signature from `existsWaiter` to `waiting` and I didn't look closely at the call site, where we were already using `coinIDString`.

JoeGruffins reviewed (commented):
Posting bond over rpc doesn't give me back a coin ID:
```
$ ./dexcctl -p abc --simnet postbond 127.0.0.1:17273 1000000000 42
{
  "bondID": "",
  "reqConfirms": 0
}
```
edit: Oh srry! I had already paid legacy fee. It works on a clean simnet client. But, I should be able to post more bond like this no? I can call multiple times if I start with bonds.

On a side note creating an dcr SPV wallet over rpc doesn't seem to be working for some reason. I'm sure unrelated.

chappjc commented on client/core/bond.go:
```diff
@@ -170,7 +338,76 @@ func (c *Core) refundExpiredBonds(ctx context.Context) {
+				continue
+			}
+
+			// For the max bonded limit, we'll normalize all bonds to the
+			// currently selected bond asset.
```
The idea of max bonded is to allow overcoming penalization, otherwise your account can get suspended (go sub zero tier) and it won't post any additional bond to get into a trading tier.

chappjc replied:
> Another thought this brings to mind is, if bond amount is changed, ideally, the user is not punished. If a full bond was bought yesterday, and the bond price goes up today, that user should not need to buy more bonds for the same tier, imo, which I believe happens here.

We can't address this easily.  Server's gonna apply the new bond amount to existing bonds when user authorizes.

chappjc commented:
> Posting bond over rpc doesn't give me back a coin ID:
> 
> ```
> $ ./dexcctl -p abc --simnet postbond 127.0.0.1:17273 1000000000 42
> {
>   "bondID": "",
>   "reqConfirms": 0
> }
> ```
> 
> edit: Oh it's working now. Not sure what was happening actually. Will try to make it happen again though.

I think you're hitting this path:

```go
		if paid {
			success = true
			// The listen goroutine is already running, now track the conn.
			c.connMtx.Lock()
			c.conns[dc.acct.host] = dc
			c.connMtx.Unlock()
			return &PostBondResult{ /* no new bond */ }, nil
		}
```

That would be existing account discovery.  Perhaps the RPC can just say ok in this case.  Not sure how else to do this one.
However, this is one of those exclusively non-UI cases because the frontend would do `DiscoverAccount` prior to `postbond`.  RPC has a `discoveraccount` route.  Perhaps that's the case you hit?

chappjc commented on client/core/bond.go:
```diff
@@ -408,23 +622,68 @@ func deriveBondKey(seed []byte, assetID, bondIndex uint32) (*secp256k1.PrivateKe
+	return keygen.GenDeepChild(seed, []uint32{hdKeyPurposeBonds})
+}
 
+func (c *Core) nextBondKey(assetID uint32) (*secp256k1.PrivateKey, error) {
```
Concept start in https://github.com/decred/dcrdex/pull/2036/commits/7522f43902d69679003cdf34645346a9849ee6d8
Will finish that idea in a sec

chappjc commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+// account tier falling below target while the replacement bond is pending. The
+// network is a parameter because expected block times are network-dependent,
+// and we require short bond lifetimes on simnet.
+func pendingBuffer(net dex.Network) int64 {
```
> I'm still on the fence here, but @chappjc makes a good point. On the timelines we're talking about, we could probably get by without asset-specific values.

Ok, I'll just do it.  Doesn't seem complex, and if we get there with ETH, the `spendableDelay` is likely to be zero (assuming we have a custom contract just for time-locking funds), so that will be helpful.

chappjc replied:
I had started on this Friday but moved on to something else because it was a PITA because I have to get the xcWallets in different spots.  Can work, just requires a bunch of annoying refactoring.

chappjc commented on client/core/core.go:
```diff
@@ -2682,6 +2687,42 @@ func (c *Core) walletIsActive(assetID uint32) bool {
+// configured DEX. includeLive should be set to consider all existing unspent
+// bonds that need to be refunded in the future.
+func (c *Core) isActiveBondAsset(assetID uint32, includeLive bool) bool {
+	for _, dc := range c.dexConnections() {
```
Hmm, started doing this with the `addressFamily` helper, but I'm not sure it it applies here. 

There are two current uses for these "is active" checks: (1) locking the wallet, and (2) disabling the wallet.

e.g. If you wanna disable the USDC wallet (not the ETH wallet), it needs to check if USDC is an "active" bond asset, but if ETH is an active bond asset we should still be able to disable the USDC wallet.

If the above doesn't work for token wallets on the basis of active trades with the base wallet, I think that's an existing bug.

chappjc replied:
Looking at child assets only though makes sense, it just appears we are looking both directions when we should only be looking to parent.

JoeGruffins commented:
> That would be existing account discovery.

Oh yeah, this was it. I can post bonds after that it seems, and now there is a message which is helpful.

JoeGruffins reviewed (approved):
Looks good to me. Will continue to test as I work on other things.

chappjc commented:
> Will continue to test as I work on other things.

Thanks for that.  Bond rotation and reserves have been challenging on account of the long durations involved.  The transaction structure and comms payloads were a breeze by comparison unfortunatley.

JoeGruffins commented:
Hmm, just had a bond seemingly disappear:
```
2023-01-31 12:21:52.336 [DBG] CORE[dcr]: 2 signature cycles to converge on fees for tx 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5: min rate = 20, actual fee rate = 20 (6420 for 321 bytes), change = true
2023-01-31 12:21:52.338 [INF] CORE: DEX 127.0.0.1:17273 has validated our bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with strength 1. 1 confirmations required to trade.
2023-01-31 12:21:52.340 [INF] CORE: Broadcasting bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST, data = 04df8bd863b17576a914807b04de9f1d58648be8564f11bddd8bb7cb352088ac.

BACKUP refund tx paying to current wallet: 0100000001d521d467a2257ec8941b6e703677113adca2bffa09d9586f5111ef1a8570171c0000000000feffffff0164b69a3b0000000000001976a9140a9f8e175d6f643315aa8db84f9742724727fad488acdf8bd863000000000100ca9a3b0000000000000000ffffffff8b473044022044c9ce7d1d23c7dac08d08393c3209ca86e3dde1e7c91fd5b233944ce92b2d4f02204960aabfe400605ce2748db668637252733b498a0a0f86284009480bd6cbb842012102114de58c61640a3caaf9e4829ffda5335c16a9a1fc956c9b21928625c9d512032004df8bd863b17576a914807b04de9f1d58648be8564f11bddd8bb7cb352088ac


2023-01-31 12:21:52.345 [INF] CORE: notify: |SUCCESS| (bondpost) BondConfirming - Waiting for 1 confirmations to post bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) to 127.0.0.1:17273
2023-01-31 12:21:58.662 [DBG] CORE[dcr]: tip change: 617 (7ea5d7933a6190e384a22a64cba0ab85351088330845244c244131c581128099) => 618 (1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df)
2023-01-31 12:21:58.662 [TRC] CORE: Processing tip change for dcr
2023-01-31 12:21:58.662 [TRC] CORE: notify: |DATA| (bondpost) RegUpdate - Fee payment confirmations 0/1
2023-01-31 12:21:59.454 [DBG] CORE[dcr]: tip change: 618 (1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df) => 619 (6213dc6370b5ad2bc0474e2604ca0c4b12d579b266bbdc9e9a41d485d13a64ff)
2023-01-31 12:21:59.454 [TRC] CORE: Processing tip change for dcr
2023-01-31 12:21:59.454 [TRC] CORE: notify: |DATA| (bondpost) RegUpdate - Fee payment confirmations 0/1
```
...
```
2023-01-31 12:29:10.320 [TRC] CORE: notify: |DATA| (bondpost) RegUpdate - Fee payment confirmations 0/1
2023-01-31 12:29:11.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 3m36s.
2023-01-31 12:29:31.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 3m16s.
2023-01-31 12:29:40.574 [DBG] CORE[dcr]: tip change: 646 (7028e2c8a01ce96ce684fb3ab36303e7544fbf49f802e9e5e46dc10cee95a168) => 647 (4f63d53fe5b0da602567f7246fce1645478e1ce600c0775e7806ea17aa9353d4)
2023-01-31 12:29:40.574 [TRC] CORE: Processing tip change for dcr
2023-01-31 12:29:40.574 [TRC] CORE: notify: |DATA| (bondpost) RegUpdate - Fee payment confirmations 0/1
2023-01-31 12:29:51.456 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 2m56s.
2023-01-31 12:30:11.456 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 2m36s.
2023-01-31 12:30:31.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 2m16s.
2023-01-31 12:30:51.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 1m56s.
2023-01-31 12:31:11.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 1m36s.
2023-01-31 12:31:31.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 1m16s.
2023-01-31 12:31:51.456 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 56s.
2023-01-31 12:32:11.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 36s.
2023-01-31 12:32:31.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) refundable in about 16s.
2023-01-31 12:32:51.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 12:33:11.458 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 12:33:31.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 12:33:51.458 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 12:34:11.458 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 12:34:31.457 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 12:34:51.458 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
```
And still not refundable but from the harness
```
./alpha getrawtransaction 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5
-5: No information available about transaction 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5
```

chappjc commented:
I've been worried about silent broadcast failures, esp. with SPV.  Really frustrating API.

I've been rotating on testnet for about 10 days, and a couple on simnet, but haven't updated my builds recently.  Will see if I broke something.

This feels a lot like the issues we have with swap redeem and refund conf checks.

Anyway, I think if you restart dexc, there's a force rebroadcast in the code.  Can you try that?  We can make that an automatic thing without restarting.

chappjc commented:
> Anyway, I think if you restart dexc, there's a force rebroadcast in the code. Can you try that? We can make that an automatic thing without restarting.

Ohhh, but I think this path only applies if it's still in `pendingBonds`, but it is now in expired...
interesting.  We should definitely rebroadcast then.

Can you check your assetdb dcr spv logs to see if there's anything suggestive of a bcast issue?

JoeGruffins commented:
Is it right:
`2023-01-31 12:21:52.336 [DBG] CORE[dcr]: 2 signature cycles to converge on fees for tx 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5: min rate = 20, actual fee rate = 20 (6420 for 321 bytes), change = true`
Are signature cycles confirmations? Should be one for this bond?

> Can you check your assetdb dcr spv logs to see if there's anything suggestive of a bcast issue?

```
2023-01-31 12:21:39.233 [INF] WLLT: Registered for transaction notifications for 209 HD address(es)
2023-01-31 12:21:39.233 [INF] WLLT: Registered for transaction notifications for all relevant outputs
2023-01-31 12:21:39.234 [INF] WLLT: Rescanning block range [1, 617]...
2023-01-31 12:21:39.243 [INF] WLLT: Rescan complete
2023-01-31 12:21:52.336 [INF] WLLT: Returning address (account=0 branch=1 child=7)
2023-01-31 12:21:52.337 [INF] WLLT: Returning address (account=0 branch=1 child=8)
2023-01-31 12:21:52.340 [INF] UDB: Inserting unconfirmed transaction 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5
2023-01-31 12:21:58.662 [INF] SPV: Connected block 1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df, height 618, 1 wallet transaction(s)
2023-01-31 12:21:59.454 [INF] SPV: Connected block 6213dc6370b5ad2bc0474e2604ca0c4b12d579b266bbdc9e9a41d485d13a64ff, height 619, 0 wallet transaction(s)
```
But from the harness, I don't see it:
```
$ ./alpha getblock 1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df
{
  "hash": "1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df",
  "confirmations": 30,
  "size": 4689,
  "height": 618,
  "version": 10,
  "merkleroot": "d400987a1ccbe08c8f7cb2d0661bc498068455ca052b195d4bb405f822f9ab61",
  "stakeroot": "7ce8d54efd7edfb7d886546a0741cf62f548a506cb03f7c1f67f81ba7f56ee70",
  "tx": [
    "df3506387aa706269f1eaacd6bda7a65f08d66a61234acb9936fcd14bac7353b",
    "7ef24034d50b1a55c4abf813064fb6a39ae992c987d27ec15b166ff34b96f31f",
    "a0eeec69dcba92ffcc087de8f5c9dc4a9a33528d9f8950e7e398c72905f12116"
  ],
  "stx": [
    "0f291d58bb523399d10e448129a241c8632c98fcc7a7c9b8969a94c9aead49b7",
    "42af15c43db46549c1dc2af4459959586ad128cdd2653d67c9befbe090e40da2",
    "d571ae4156127664daf01fd7df27398a8cf673d7014b916a069067f9734be51d",
    "bec39dba0d79bbeab6b4959db3a52a4d2998c034c19ed6047d0b848b7e2e014c",
    "298ce5b3572027c2fa749eb129aa8b5fe7f314cd3f316eaa7358a4050b51d551",
    "d36db760795b37d4bdc79f8cdfebcb65d0814e1f4b75496a7498923ac23ae8a5",
    "3087b37deaa66dcca8098c96c107bfdd9e87a06c317ff85c16a92757df766af5",
    "0540e2c4544dd48b23c7bdbc8c4d8143b1d1acfe6f2b431393a9bfee674e3c4d",
    "574cd00c6b9cd06ddca86dfbede7077c589b2b34ac1742f70f64227656db20eb",
    "a991c2f662ada50ff9236fb486f059b82a9d30f0950b3cc4f3597c5b5bb195b8",
    "97dbbf5f31444d50c97191d0d9949bd903f2fa3b9885bfddba55e0efe757d6ec",
    "e55a310e9e03aac947235e8b7f299e47f139ff5d8931a5c58c120b88fe1f9ef0"
  ],
  "time": 1675135318,
  "mediantime": 1675134672,
  "nonce": 0,
  "votebits": 1,
  "finalstate": "e17a52b583a0",
  "voters": 5,
  "freshstake": 6,
  "revocations": 0,
  "poolsize": 145,
  "bits": "207fffff",
  "sbits": 0.00542888,
  "extradata": "484d15015b79c295000000000000000000000000000000000000000000000000",
  "stakeversion": 9,
  "difficulty": 1,
  "chainwork": "00000000000000000000000000000000000000000000000000000000000004d6",
  "previousblockhash": "7ea5d7933a6190e384a22a64cba0ab85351088330845244c244131c581128099",
  "nextblockhash": "6213dc6370b5ad2bc0474e2604ca0c4b12d579b266bbdc9e9a41d485d13a64ff"
}
```

chappjc commented:
> Is it right: `2023-01-31 12:21:52.336 [DBG] CORE[dcr]: 2 signature cycles to converge on fees for tx 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5: min rate = 20, actual fee rate = 20 (6420 for 321 bytes), change = true` Are signature cycles confirmations? Should be one for this bond?

That's fine.  It's just how many times it had to iterate on the fees.

> 2023-01-31 12:21:52.336 [INF] WLLT: Returning address (account=0 branch=1 child=7)
> 2023-01-31 12:21:52.337 [INF] WLLT: Returning address (account=0 branch=1 child=8)
> 2023-01-31 12:21:52.340 [INF] UDB: Inserting unconfirmed transaction 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5
> 2023-01-31 12:21:58.662 [INF] SPV: Connected block 1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df, height 618, 1 wallet transaction(s)
> 2023-01-31 12:21:59.454 [INF] SPV: Connected block 6213dc6370b5ad2bc0474e2604ca0c4b12d579b266bbdc9e9a41d485d13a64ff, height 619, 0 wallet transaction(s)

wallet thinks it was mined in that block, but node disagrees...  eh, did our simnet nodes fork?

JoeGruffins commented:
Oh, heres something from dcrd logs:
```
2023-01-31 12:21:52.343 [DBG] PEER: Received inv (tx 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5) from 127.0.0.1:40272 (inbound)
2023-01-31 12:21:52.343 [DBG] PEER: Sending getdata (tx 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5) to 127.0.0.1:40272 (inbound)
2023-01-31 12:21:52.343 [DBG] PEER: Adding deadline for command getdata for peer 127.0.0.1:40272
2023-01-31 12:21:52.343 [DBG] PEER: Received tx (hash 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5, 1 inputs, 3 outputs, lock height 0) from 127.0.0.1:40272 (inbound)
2023-01-31 12:21:52.343 [DBG] SYNC: Rejected transaction 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5 from 127.0.0.1:40272 (inbound): transaction a0eeec69dcba92ffcc087de8f5c9dc4a9a33528d9f8950e7e398c72905f12116 in the pool already spends the same coins
```

chappjc commented:
Beautiful.  If you have the nodes up or the logs available, pls try to figure out what the prior transaction that is supposedly spending the same coins.

JoeGruffins commented:
<details closed>
<summary>tx we are trying to double spend</summary>

```
[joe@hyrule harness-ctl]$ ./alpha getrawtransaction a0eeec69dcba92ffcc087de8f5c9dc4a9a33528d9f8950e7e398c72905f12116 1
{
  "hex": "01000000017bc7491b99f38bfc268a9ca609ba78dab40c99ce5ab2490fe9fb695b2e6b3e320100000000ffffffff0300ca9a3b00000000000017a914ec440eceed7d98392391cc5196305ba6570d799b87000000000000000000003c6a3a0000289a60428c59174bb7d5ace52ebe6119d975b63ec507c33a8b1496ce97386d2d63d8897d807b04de9f1d58648be8564f11bddd8bb7cb3520daa1aa5f0100000000001976a914200047b4a5768d91c1a6923564424cd5f8ef41d588ac0000000000000000015a78459b0100000064020000020000006a47304402203a62bde80af76c11acc93ae22a29b324777ce1a57cacbaad1a252099b6e9744f02204808497a27311ede240c351eb0da593602078f4b91440052d7bfc2f8e609629101210340f05f9b419c4ec766ff70d592b79944b0314e5bb86d52f04f3cfa38c5fa4907",
  "txid": "a0eeec69dcba92ffcc087de8f5c9dc4a9a33528d9f8950e7e398c72905f12116",
  "version": 1,
  "locktime": 0,
  "expiry": 0,
  "vin": [
    {
      "txid": "323e6b2e5b69fbe90f49b25ace990cb4da78ba09a69c8a26fc8bf3991b49c77b",
      "vout": 1,
      "tree": 0,
      "sequence": 4294967295,
      "amountin": 68.9998857,
      "blockheight": 612,
      "blockindex": 2,
      "scriptSig": {
        "asm": "304402203a62bde80af76c11acc93ae22a29b324777ce1a57cacbaad1a252099b6e9744f02204808497a27311ede240c351eb0da593602078f4b91440052d7bfc2f8e609629101 0340f05f9b419c4ec766ff70d592b79944b0314e5bb86d52f04f3cfa38c5fa4907",
        "hex": "47304402203a62bde80af76c11acc93ae22a29b324777ce1a57cacbaad1a252099b6e9744f02204808497a27311ede240c351eb0da593602078f4b91440052d7bfc2f8e609629101210340f05f9b419c4ec766ff70d592b79944b0314e5bb86d52f04f3cfa38c5fa4907"
      }
    }
  ],
  "vout": [
    {
      "value": 10,
      "n": 0,
      "version": 0,
      "scriptPubKey": {
        "asm": "OP_HASH160 ec440eceed7d98392391cc5196305ba6570d799b OP_EQUAL",
        "hex": "a914ec440eceed7d98392391cc5196305ba6570d799b87",
        "reqSigs": 1,
        "type": "scripthash",
        "addresses": [
          "ScxPDFvHbi4myd99kMpe9n5tGYLKesoQxCh"
        ],
        "version": 0
      }
    },
    {
      "value": 0,
      "n": 1,
      "version": 0,
      "scriptPubKey": {
        "asm": "OP_RETURN 0000289a60428c59174bb7d5ace52ebe6119d975b63ec507c33a8b1496ce97386d2d63d8897d807b04de9f1d58648be8564f11bddd8bb7cb3520",
        "hex": "6a3a0000289a60428c59174bb7d5ace52ebe6119d975b63ec507c33a8b1496ce97386d2d63d8897d807b04de9f1d58648be8564f11bddd8bb7cb3520",
        "type": "nulldata",
        "version": 0
      }
    },
    {
      "value": 58.9998537,
      "n": 2,
      "version": 0,
      "scriptPubKey": {
        "asm": "OP_DUP OP_HASH160 200047b4a5768d91c1a6923564424cd5f8ef41d5 OP_EQUALVERIFY OP_CHECKSIG",
        "hex": "76a914200047b4a5768d91c1a6923564424cd5f8ef41d588ac",
        "reqSigs": 1,
        "type": "pubkeyhash",
        "addresses": [
          "SsXGUeXKfh4CUHHZA4HpXTBTMwtWHpEbjum"
        ],
        "version": 0
      }
    }
  ],
  "blockhash": "1c5f0107737e90e79a2f5a72313ad75ed003a9ea898ae9eb8cdd85696805a2df",
  "blockheight": 618,
  "blockindex": 2,
  "confirmations": 30,
  "time": 1675135318,
  "blocktime": 1675135318
}
```
</details>

chappjc commented:
OK, so another bond.  Will investigate shortly.  Not sure how that happened, but it feels like a coin locking issue.

JoeGruffins commented:
You know, maybe I deleted the db, and this is from a new node created from the same seed that apparently did not find the bond tx from before.

JoeGruffins commented:
`a0eeec69dcba92ffcc087de8f5c9dc4a9a33528d9f8950e7e398c72905f12116` is in the block after we try to send the doublespend, so likely it was a tx in mempool that the new spv wallet did not see until mined.

JoeGruffins commented:
<details closed>
<summary>Starting back up dexc logs</summary>

```
2023-01-31 13:52:29.995 [INF] CORE: Loaded expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0. Refund tx: 0100000001d521d467a2257ec8941b6e703677113adca2bffa09d9586f5111ef1a8570171c0000000000feffffff0164b69a3b0000000000001976a9140a9f8e175d6f643315aa8db84f9742724727fad488acdf8bd863000000000100ca9a3b0000000000000000ffffffff8b473044022044c9ce7d1d23c7dac08d08393c3209ca86e3dde1e7c91fd5b233944ce92b2d4f02204960aabfe400605ce2748db668637252733b498a0a0f86284009480bd6cbb842012102114de58c61640a3caaf9e4829ffda5335c16a9a1fc956c9b21928625c9d512032004df8bd863b17576a914807b04de9f1d58648be8564f11bddd8bb7cb352088ac
```
...
```
2023-01-31 13:52:49.996 [INF] CORE: Connecting wallet for dcr
2023-01-31 13:52:50.007 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
```
...
```
2023-01-31 13:52:59.118 [DBG] CORE: Preparing dcr wallet to maintain target tier of 1 for 127.0.0.1:17273, bonding limit 0
2023-01-31 13:52:59.205 [ERR] CORE: notify: |ERROR| (dex_auth) DEX auth error - 127.0.0.1:17273: 'connect' error: rpc error: error code 50: no account found for account ID: 809e7bbdca365e928e24f08bd981f61e2a9b5c957d8e8b28c5ab4d6a718b9376
```
...
```
2023-01-31 13:53:09.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:53:29.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:53:49.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:54:09.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:54:29.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:54:49.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:55:09.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
2023-01-31 13:55:29.996 [DBG] CORE: Expired bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) with lock time 2023-01-31 12:32:47 +0900 JST not yet refundable according to wallet.
```
```
$ date
Tue Jan 31 01:56:05 PM JST 2023
```
<details>

So, it this very weird circumstance that would be difficult to accomplish on mainnet, I'm kinda stuck with the nonexistent bond.

JoeGruffins commented:
Oh, I lied, just needed to progress the chain:
```
2023-01-31 14:01:50.002 [INF] CORE: notify: |SUCCESS| (bondrefund) BondRefunded - Bond 1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:0 (dcr) for 127.0.0.1:17273 refunded in f11156d2cafb9ff7a3c6f4f8aea33becf6e70039b694fa0cfbaa07e53758a488:0 (dcr)
```

JoeGruffins commented:
I posted a new bond and am now able to trade, so the problem was resolved eventually.

chappjc commented:
I guess I should have mentioned, that with simnet and the locktimes and fast rates, you really need `watch -n 15 ./mine-alpha 1` running constantly. It all falls apart otherwise.

JoeGruffins commented:
Had it happen again, this time just running normally. May be a result of fast simnet things...

dexc:
```
2023-01-31 14:47:48.650 [INF] CORE: DEX 127.0.0.1:17273 has validated our bond 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df:0 (dcr) with strength 3. 1 confirmations required to trade.
2023-01-31 14:47:48.653 [INF] CORE: Broadcasting bond 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df:0 (dcr) with lock time 2023-01-31 14:58:43 +0900 JST, data = 0413aed863b17576a9148f720461f71c516fb0ae82db88dcd52927f9c59388ac.

BACKUP refund tx paying to current wallet: 0100000001df97a74e2ea2e5818540fa66102017e4ffdb62d1fcc4d6bdc62b4f4eae828c880000000000feffffff013254d0b20000000000001976a9143fd38f097244f38ecd8fa231c22eb4b0d92e0a4f88ac13aed8630000000001005ed0b20000000000000000ffffffff8b47304402207a4cf79269474ce0a56d0342b7c033b13b32670eb3db84231a01189cf10173af02201d38108c395723509abb92f570aeead1deb5cad4a49050edb57e015d49a898b101210273bdf8da5e8b3a83b0d27026bc248de7ca891af9c2ede863c36e394d64b3cc18200413aed863b17576a9148f720461f71c516fb0ae82db88dcd52927f9c59388ac


2023-01-31 14:47:48.656 [INF] CORE: notify: |SUCCESS| (bondpost) BondConfirming - Waiting for 1 confirmations to post bond 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df:0 (dcr) to 127.0.0.1:17273
```
dcrd:
```
2023-01-31 14:47:48.655 [DBG] PEER: Received inv (tx 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df) from 127.0.0.1:56364 (inbound)
2023-01-31 14:47:48.655 [DBG] PEER: Sending getdata (tx 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df) to 127.0.0.1:56364 (inbound)
2023-01-31 14:47:48.655 [DBG] PEER: Adding deadline for command getdata for peer 127.0.0.1:56364
2023-01-31 14:47:48.655 [DBG] PEER: Received tx (hash 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df, 4 inputs, 3 outputs, lock height 0) from 127.0.0.1:56364 (inbound)
2023-01-31 14:47:48.655 [DBG] TXMP: Expired 1 orphan (remaining: 0)
2023-01-31 14:47:48.655 [DBG] TXMP: Stored orphan transaction 888c82ae4e4f2bc6bdd6c4fcd162dbffe417201066fa408581e5a22e4ea797df (total: 1)
```

JoeGruffins commented:
And again, with the same `orphan transaction` in dcrd.

chappjc commented:
I don't know what to make of "orphan transaction" from dcrd. Are the simnet
nodes reporting reorgs?

What kind of dcr wallet is dexc using? Native or RPC?  I actually haven't
tried external dcrwallet.


On Tue, Jan 31, 2023, 12:05 AM JoeGruffins ***@***.***> wrote:

> And again, with the same orphan transaction in dcrd.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/decred/dcrdex/pull/2036#issuecomment-1409814176>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHQOSIXUXHKW3LEYX436ADWVCTSXANCNFSM6AAAAAATW3NCYE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JoeGruffins commented:
> Are the simnet
nodes reporting reorgs?

no

> What kind of dcr wallet is dexc using?

internal wallet

I bet it is trying to spend that original `1c1770851aef11516f58d909fabfa2dc3a117736706e1b94c87e25a267d421d5:#` it thinks exists.

JoeGruffins commented:
Deleting the db and recreating from seed seems to have stopped the bad transactions. I'm pretty sure it was that initial bad tx, the change being "spent".

chappjc commented on client/core/bond.go:
```diff
@@ -39,58 +40,225 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+			}
+			lockTimeThresh += int64(cfg.BondExpiry)
 		}
-		dc.cfgMtx.RUnlock()
+		replaceThresh := lockTimeThresh + pendingBuffer(c.net) // replace before expiry to avoid tier drop
```
Nah, we just have to be at this spot to process refunds.  I'll double check that all the logic below excludes never connected dex (cfg == nil)

buck54321 replied:
`lockTimeThresh` and `replaceThresh` would be too early, which I guess means bonds **wouldn't** be expired or replaced when maybe they should be, so limited damage, but I'm not feeling great about running `filterExpiredBonds` with bad numbers.

chappjc replied:
> which I guess means bonds **wouldn't** be expired or replaced when maybe they should be, so limited damage

Right, this is totally intended though.  In the absence of expiry info, don't move bonds into the expired bucket until we actually reach locktime. A few lines up this is doc'd:

``` go
lockTimeThresh := now // in case dex is down, expire (to refund when lock time is passed)
```

`replaceThresh` only affects creation of the `weak` slice, which is purely local to this method, no fields of `dc.acct` modified in doing that.  And if all important guards on the posting branch block any action:

```go
if mustPost > 0 && targetTier > 0 && bondExpiry > 0 {
	c.log.Infof("Gotta post %d bond increments now. Target tier %d, current tier %d (%d weak, %d pending)",
		mustPost, targetTier, tier, weakStrength, pendingStrength)
	if !unlocked || dc.status() != comms.Connected {
		c.log.Warnf("Unable to post the required bond while disconnected or logged out.")
		continue
	}
```

Here the absence of a `cfg` is implied by both of `bondExpiry` and `unlocked`, with `dc.status` being strongly correlated.

Happy to introduce a new flag to be extra safe though.

chappjc commented on client/core/bond.go:
```diff
@@ -39,58 +40,217 @@ func (c *Core) watchExpiredBonds(ctx context.Context) {
+//  - when creating a bond, set lockTime = 2*(BondExpiry+pendingBuffer)+spendableDelay
+//  - create a replacement bond at lockTime-BondExpiry-pendingBuffer
+
+// pendingBuffer gives the duration in seconds prior to reaching bond expiry
+// (account, not lockTime) after which a new bond should be posted to avoid
+// account tier falling below target while the replacement bond is pending. The
+// network is a parameter because expected block times are network-dependent,
+// and we require short bond lifetimes on simnet.
+func pendingBuffer(net dex.Network) int64 {
```
Closely related concepts, but this is a high buffer based on what you've described as `estimatedSecsToBondConf`, where a _very large percentage_ of real life time-to-confirm would be covered by `pendingBuffer`.  I will allude to that in a revision to the comment.

chappjc commented on client/core/bond.go:
```diff
@@ -123,11 +291,21 @@ func (c *Core) refundExpiredBonds(ctx context.Context) {
 			// Generate a refund tx paying to an address from the currently
 			// connected wallet, using bond.PrivKey to create the signed
 			// transaction. The RefundTx is really a backup.
 			priv := secp256k1.PrivKeyFromBytes(bond.PrivKey)
+			// c.bondKeyIdx(bond.AssetID, bond.KeyIndex) // TODO, with KeyIndex in DB instead of PrivKey
```
grr dang I forgot about this KeyIndex
We'll have to toast our DBs unless I do it now.

chappjc replied:
I mean we'll have to toast them anyway if we were using this PR, but I'd hate to put a bad DB format on master.  Sec...

chappjc replied:
Ugh, the `db.Bond` structure on master has `BondPrivKey`.
I'll put some minimal handling for that interim struct version so we don't have to nuke our testnet and mainnet DBs that we may have used on master.  Shame to have perpetual edge case code for unreleased revision handling... let's see how it looks.

chappjc commented:
Added a wallet synced gate because I believe I was able to (automatically) post a bond tx that double spent coins: https://github.com/decred/dcrdex/pull/2036/commits/ca9914c206b7ce5238280fa894c850945cbba71d

For roughly 6 days I had been using a different copy of the data dir, including the DCR native wallet, to work with ETH/ERC20 swaps, and when I came back to this, it tried to post new bond to hit target tier before the wallet had synced.  The DCR wallet happily funded the bond txns with spent coins because it was still scanning blocks.

Also added rebroadcasting on certain unusual paths to `monitorBondConfs`.
